checker/check_result.go

```diff
@@ -16,30 +16,178 @@ package checker
 
 import (
 	"errors"
+	"fmt"
+	"math"
 
 	scorecarderrors "github.com/ossf/scorecard/errors"
 )
 
-const MaxResultConfidence = 10
+// UPGRADEv2: to remove.
+const (
+	MaxResultConfidence  = 10
+	HalfResultConfidence = 5
+	MinResultConfidence  = 0
+)
 
 // ErrorDemoninatorZero indicates the denominator for a proportional result is 0.
+// UPGRADEv2: to remove.
 var ErrorDemoninatorZero = errors.New("internal error: denominator is 0")
 
+//nolint
+const (
+	MaxResultScore          = 10
+	HalfResultScore         = 5
+	MinResultScore          = 0
+	InconclusiveResultScore = -1
+)
+
+//nolint
 type CheckResult struct {
+	// Old structure
 	Error       error `json:"-"`
 	Name        string
 	Details     []string
 	Confidence  int
 	Pass        bool
 	ShouldRetry bool `json:"-"`
+
+	// UPGRADEv2: New structure. Omitting unchanged Name field
+	// for simplicity.
+	Version  int           `json:"-"` // Default value of 0 indicates old structure.
+	Error2   error         `json:"-"` // Runtime error indicate a filure to run the check.
+	Details2 []CheckDetail `json:"-"` // Details of tests and sub-checks
+	Score2   int           `json:"-"` // {[0...1], -1 = Inconclusive}
+	Reason2  string        `json:"-"` // A sentence describing the check result (score, etc)
+}
+
+// CreateResultWithScore is used when
+// the check runs without runtime errors and we want to assign a
+// specific score.
+func CreateResultWithScore(name, reason string, score int) CheckResult {
+	pass := true
+	//nolint
+	if score < 8 {
+		pass = false
```

Comment:
put 8 in a const somewhere? it's not clear exactly otherwise what this is. Same in other places (like line 97)

and maybe also just something like

```go
pass := score >= PassThreshold
```